web/src/features/hfiCalculator/slices/stationsSlice.ts

```diff
@@ -0,0 +1,113 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit'
+
+import { WeatherStation, PlanningArea, FireCentre, getHFIStations } from 'api/hfiCalcAPI'
+import { AppThunk } from 'app/store'
+import { logError } from 'utils/error'
+
+interface State {
+  loading: boolean
+  error: string | null
+  weatherStations: WeatherStation[]
+  fireCentres: FireCentre[]
+  planningAreasByFireCentre: Record<string, PlanningArea[] | undefined>
+  stationsByPlanningArea: Record<string, WeatherStation[] | undefined>
+}
+
+const initialState: State = {
+  loading: false,
+  error: null,
+  weatherStations: [],
+  fireCentres: [],
+  planningAreasByFireCentre: {},
+  stationsByPlanningArea: {}
+}
+
+const stationsSlice = createSlice({
+  name: 'hfiStations',
+  initialState,
+  reducers: {
+    getHFIStationsStart(state: State) {
+      state.error = null
+      state.loading = true
+      state.weatherStations = []
+      state.fireCentres = []
+      state.planningAreasByFireCentre = {}
+      state.stationsByPlanningArea = {}
+    },
+    getHFIStationsFailed(state: State, action: PayloadAction<string>) {
+      state.error = action.payload
+      state.loading = false
+    },
+    getHFIStationsSuccess(state: State, action: PayloadAction<WeatherStation[]>) {
+      state.error = null
+      state.weatherStations = action.payload
+      action.payload.forEach(station => {
+        // if the station's fire centre isn't already in state.fireCentres array, add it
+        if (
```

Comment:
You can simplify these linear searches by using a set that will do this for you in constant time: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

Reply:
Or do it on the backend like Sy suggested: https://github.com/bcgov/wps/pull/1045#discussion_r651306248
Probably easier to test and satisfy the redux dispatch interfaces that way.